locksmithd: ignore nil sent by dbus.Signal

rfairley opened the pull request:
Specifies the loop in RebootNeededSignal to continue if a 'nil' signal is
sent to the client channel connected to the D-Bus. This avoids
dereferencing a nil datatype in NewStatus, which causes the panic in
coreos/bugs#1987.

fixes coreos/bugs#1987

Commit:
locksmithd: continue on dbus.Signal of nil or incorrect type

Specifies the loop in RebootNeededSignal to continue if a 'nil' signal
or a signal with fields of incorrect type is sent. This avoids a panic
on a nil dereference (seen in coreos/bugs#1987) or a failed type assertion.

fixes coreos/bugs#1987

rfairley commented on updateengine/client.go:
```diff
@@ -63,7 +63,7 @@ func New() (c *Client, err error) {
 	}
 
-	c.object = c.conn.Object("com.coreos.update1", dbus.ObjectPath(dbusPath))
+	c.object = c.conn.Object("com.coreos.update1", dbus.ObjectPath(dbusPath)).(*dbus.Object)
```
Checked on my machine and this seems to be the line breaking the build - will edit tomorrow.

lucab replied:
If you think this related to different go toolchain version, feel free to add 1.9.x and 1.10.x to travis. Please note that locksmith relies on an old vendored copy of go-dbus for this, so double-check your environment and dependency resolution first.

lucab replied:
If you think this related to different go toolchain versions, feel free to add 1.9.x and 1.10.x to travis. Please note that locksmith relies on an old vendored copy of go-dbus for this, so double-check your environment and dependency resolution first.

lucab commented on updateengine/status.go:
```diff
@@ -42,6 +42,32 @@ func NewStatus(body []interface{}) (s Status) {
+
+	// Check that elements of body have correct type corresponding to type Status
+	typeIsCorrect := false
+	for i, elem := range body {
```
This code is somewhat awkward. I would suggest instead to move the nil+length+types checking directly into a fallible constructor (e.g. `TryNewStatus` or similar), as this file is the only consumer.
It would look like `TryNewStatus(*dbus.Signal) (Status, error)` so that you can `continue` on error in the loop above.

rfairley replied:
I see - makes sense to make use of `error`. Will work on this now. Thank you!

Commit:
.travis.yml: update go version to 1.10.x

Adds go versions 1.10.x and 1.9.x to the Travis CI build. Removes
older versions 1.8.x and 1.7.x and keeps a minimum known working
version 1.6.x.

lucab commented on updateengine/status.go:
```diff
@@ -28,6 +31,56 @@ type Status struct {
+	errStatusNil    = errors.New("dbus signal is nil")
+	errStatusLength = errors.New("signal body has insufficient length")
+	errStatusType   = errors.New("signal body has incorrect type")
+	statusZero      = Status{0, 0, "", "", 0}
```
No need for this here. You can either move it to be local to the function, or even better just `return Status{}, errFoo`.

rfairley replied:
Made an edit - how's this?

rfairley commented:
Done! Please let me know if there is anything else. Otherwise, I think it is good to merge.

lucab commented:
/cc @sdemos @arithx in case you want a second pass before merging this. We also need to update the maintainers file at some point.

arithx commented on updateengine/status.go:
```diff
@@ -28,6 +31,55 @@ type Status struct {
+// than required, or has a field of incorrect type (types must be int64, float64,
+// string, string, int64 respectively). If successful, the new status is returned.
+func TryNewStatus(signal *dbus.Signal) (Status, error) {
+
```
extra line break

rfairley replied:
Fixing that, thanks!

arithx commented on updateengine/status.go:
```diff
@@ -28,6 +31,55 @@ type Status struct {
+var (
+	errStatusNil    = errors.New("dbus signal is nil")
+	errStatusLength = errors.New("signal body has insufficient length")
+	errStatusType   = errors.New("signal body has incorrect type")
```
@lucab: do you think it would be worthwhile to format this error with the failing field?

lucab replied:
At the moment it is not really used, so probably not. If we need to surface that somewhere later, we can quickly do there.

rfairley replied:
Will leave it for now. Would the failing field look something like `locksmith: error message here`?

lucab replied:
@rfairley it would be something like `fmt.Errorf("signal field %q has incorrect type", fieldname)` in `TryNewStatus` type checks.

rfairley replied:
Right, I see. I think it can be left without specifying the field - I agree it can be resurfaced quickly if the need comes up.

arithx reviewed (approved):
Kola run through DO looks good.

ajeddeloh commented on updateengine/client_test.go:
```diff
@@ -93,6 +93,20 @@ func TestRebootNeededSignal(t *testing.T) {
+	time.Sleep(10 * time.Millisecond)
+
+	select {
+	case stat := <-r:
```
Do we want to log anything here? Otherwise the error is just eaten?

rfairley replied:
Is this logging for the test only? It would better to log something for the test case of sending the `nil` signal. (right now `"some other ignored signal"` gets a status printed for it when running `go test` in locksmith/updateengine).

I agree on not logging `RebootNeededSignal` in the case that the loop `continue`s as this would pollute the shutdown logs. Logging one time only if `continue` happens by means of a counter or flag in the `for` loop would be OK though since `nil` signals don't always happen, and it might be useful to know from the shutdown logs if it did. We could add giving one log message for the first failure of `TryNewStatus` in, WDYT?

On logging for the test (in `client_test.go` a log should be added to indicate that the `nil` signal test is happening. I can add this in a bit.

ajeddeloh replied:
sgtm, and yeah I meant to comment on the continue in the code, not on this test.

rfairley replied:
Sounds good, thanks.

On `client_test.go`: Had a second thought (from my comment above) on logging in `client_test.go`, and it is OK as it is since `t.Fatal` is called in case of failure on the `nil` signal test. So `client_test.go` will be kept as in this PR, unless there is anything else to add.

On `RebootNeededSignal`: We can still add a log for the first failure only (so logging the failure happens only once) of `TryNewStatus` in `RebootNeededSignal`, if we want to log *something* for the `nil` signal showing without polluting logs. However it may just be adding unnecessary logic to the code.

If all is good as the PR is now, I think it is OK to merge.

ajeddeloh reviewed (approved):
LGTM

lucab commented:
@ajeddeloh @arithx fine to merge then?

ajeddeloh commented:
good to merge (idk if you have merge access?)

rfairley commented:
Good to merge.

lucab commented:
@rfairley we need to cut a release with this fix and then forward it to ContainerLinux. Would you mind preparing a PR for the 0.6.2 release? It should pretty much look like this (with a relevant changelog entry): https://github.com/coreos/locksmith/pull/134

rfairley commented:
@lucab Done - opened PR #138
